File: leetcode/binary_tree_level_order_traversal/binary_tree_level_order_traversal.java
package binary_tree_level_order_traversal;

import java.util.ArrayList;
import java.util.List;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> levelOrder(TreeNode root) {
        traverse(root, 0);
        return ans;
    }

    private void traverse(TreeNode root, int level) {
        if(root == null) return;
        if(ans.size() == level) {
            List<Integer> tmp = new ArrayList<>();
            tmp.add(root.val);
            ans.add(tmp);
        } else {
            ans.get(level).add(root.val);
        }

        level++;
        traverse(root.left, level);
        traverse(root.right, level);
    }
}

File: leetcode/binary_tree_level_order_traversal/binary_tree_level_order_traversal.cpp
/*
 * Time Complexity: O(N)
 *   - N is the number of nodes in the tree.
 *   - Each node is visited once.
 *
 * Space Complexity: O(N)
 *   - For the answer vector and recursion stack (in the worst case).
 */

#include <vector>

using std::vector;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        traverse(root, ans, 0);
        return ans;
    }

    void traverse(TreeNode* root, vector<vector<int>>& ans, int level) {
        if(root == nullptr) return;
        if(ans.size() == level) {
            ans.push_back({root->val});
        } else {
            ans[level].push_back(root->val);
        }

        level++;
        traverse(root->left, ans, level);
        traverse(root->right, ans, level);
    }
};